src/OrbitGl/App.cpp

```diff
@@ -411,35 +411,67 @@ void OrbitApp::OnTimer(const TimerInfo& timer_info) {
 
 void OrbitApp::OnMemoryEventWrapper(
     const orbit_grpc_protos::MemoryEventWrapper& memory_event_wrapper) {
-  if (!memory_event_wrapper.has_system_memory_usage()) return;
-
-  const orbit_grpc_protos::SystemMemoryUsage& system_memory_usage =
-      memory_event_wrapper.system_memory_usage();
-  TimerInfo timer_info;
-  timer_info.set_type(TimerInfo::kSystemMemoryUsage);
-  timer_info.set_start(system_memory_usage.timestamp_ns());
-  timer_info.set_end(system_memory_usage.timestamp_ns());
-
-  // The RepeatedField<Element>::Set(int index, const Element & value) method won't update the
-  // current_size_ of the repeated field. If the current_size_ == 0, the repeated field will be
-  // ignored when copying the protobuf. Therefore, if encoding the following memory information
-  // into TimerInfo registers with the Set method, the registers field of the copied TimerInfo
-  // would be empty.
-  std::vector<uint64_t> encoded_values(static_cast<size_t>(SystemMemoryUsageEncodingIndex::kEnd));
-  encoded_values[static_cast<size_t>(SystemMemoryUsageEncodingIndex::kTotalKb)] =
-      orbit_api::Encode<uint64_t>(system_memory_usage.total_kb());
-  encoded_values[static_cast<size_t>(SystemMemoryUsageEncodingIndex::kFreeKb)] =
-      orbit_api::Encode<uint64_t>(system_memory_usage.free_kb());
-  encoded_values[static_cast<size_t>(SystemMemoryUsageEncodingIndex::kAvailableKb)] =
-      orbit_api::Encode<uint64_t>(system_memory_usage.available_kb());
-  encoded_values[static_cast<size_t>(SystemMemoryUsageEncodingIndex::kBuffersKb)] =
-      orbit_api::Encode<uint64_t>(system_memory_usage.buffers_kb());
-  encoded_values[static_cast<size_t>(SystemMemoryUsageEncodingIndex::kCachedKb)] =
-      orbit_api::Encode<uint64_t>(system_memory_usage.cached_kb());
-
-  *timer_info.mutable_registers() = {encoded_values.begin(), encoded_values.end()};
-
-  GetMutableTimeGraph()->ProcessTimer(timer_info, nullptr);
+  if (memory_event_wrapper.has_system_memory_usage()) {
```

Comment:
Nit: Have you considered making separate functions for processing system and cgroup memory usage? This way we would avoid these large if-branches.

Reply:
Thanks for the suggestion. Now they are put into separate functions, i.e., `CaptureEventProcessorForListener::ExtractAndProcessSystemMemoryTrackingTimer` and `CaptureEventProcessorForListener::ExtractAndProcessCGroupAndProcessMemoryTrackingTimer` in `CaptureEventProcessor.cpp`.